indexing-service/src/main/java/io/druid/indexing/common/task/ArchiveTask.java

```diff
@@ -91,12 +98,15 @@ public TaskStatus run(TaskToolbox toolbox) throws Exception
       log.info("OK to archive segment: %s", unusedSegment.getIdentifier());
     }
 
-    // Move segments
-    for (DataSegment segment : unusedSegments) {
-      final DataSegment archivedSegment = toolbox.getDataSegmentArchiver().archive(segment);
-      toolbox.getTaskActionClient().submit(new SegmentMetadataUpdateAction(ImmutableSet.of(archivedSegment)));
+    if (toolbox.getTaskActionClient().submit(new LockUpgradeAction(getInterval()))) {
+      // Move segments
+      for (DataSegment segment : unusedSegments) {
+        final DataSegment archivedSegment = toolbox.getDataSegmentArchiver().archive(segment);
```

Comment:
This is a lot of work to be doing inside a non-preemptible block; it can take quite a long time to archive of a lot of segments. maybe we should chunk things up so it releases and re-acquires the lock every N archives?